model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala

```diff
@@ -24,69 +23,172 @@ object ArrowInliner extends Logging {
   ): State[S, OpModel.Tree] =
     callArrowRet(arrow, call).map(_._1)
 
+  // Get streams that was declared outside of a function
+  private def getOutsideStreamNames[S: Exports]: State[S, Set[String]] =
+    Exports[S].exports
+      .map(exports =>
+        exports.collect { case (n, VarModel(_, StreamType(_), _)) =>
+          n
+        }.toSet
+      )
+
+  // push results to streams if they are exported to streams
+  private def pushStreamResults[S: Mangler: Exports: Arrows](
+    outsideStreamNames: Set[String],
+    exportTo: List[CallModel.Export],
+    results: List[ValueRaw],
+    body: OpModel.Tree
+  ): State[S, (List[OpModel.Tree], List[ValueModel])] =
+    for {
+      // Fix return values with exports collected in the body
+      resolvedResult <- RawValueInliner.valueListToModel(results)
+    } yield {
+      // Fix the return values
+      (exportTo zip resolvedResult)
+        .map {
+          case (
+                CallModel.Export(n, StreamType(_)),
+                (res @ VarModel(_, StreamType(_), _), resDesugar)
+              ) if !outsideStreamNames.contains(n) =>
+            resDesugar.toList -> res
+          case (CallModel.Export(exp, st @ StreamType(_)), (res, resDesugar)) =>
+            // pass nested function results to a stream
+            (resDesugar.toList :+ PushToStreamModel(
+              res,
+              CallModel.Export(exp, st)
+            ).leaf) -> VarModel(
+              exp,
+              st,
+              Chain.empty
+            )
+          case (_, (res, resDesugar)) =>
+            resDesugar.toList -> res
+        }
+        .foldLeft[(List[OpModel.Tree], List[ValueModel])](
+          (body :: Nil, Nil)
+        ) { case ((ops, rets), (fo, r)) =>
+          (fo ::: ops, r :: rets)
+        }
```

Comment:
I suggest using `.unzip.leftMap(_.flatten)` instead of this (with `import cats.syntax.bifunctor.*`)
Also `body` is not used anywhere in this function but here. Lets just add `body` in `inline` and not pass it to this function.